samples/sync_object.html

```diff
@@ -0,0 +1,422 @@
+        }
+
+        function emitParticles() {
+            time = Date.now();
+
+            var programEmit = programs[PROGRAM_EMIT];
+
+            // Toggle source and destination VBO
+            var sourceVBO = particleVBOs[currentSourceIdx];
+            var sourceVAO = particleVAOs[currentSourceIdx];
+
+            var destinationVBO = particleVBOs[(currentSourceIdx + 1) % 2];
+
+            gl.bindTransformFeedback(gl.TRANSFORM_FEEDBACK, transformFeedback);
+
+            gl.useProgram(programEmit);
+
+            setupVertexAttributes(sourceVAO, sourceVBO);
+
+            // Set transform feedback buffer
+            gl.bindBufferBase(gl.TRANSFORM_FEEDBACK_BUFFER, 0, destinationVBO[Particle.POSITION]);
+            gl.bindBufferBase(gl.TRANSFORM_FEEDBACK_BUFFER, 1, destinationVBO[Particle.VELOCITY]);
+            gl.bindBufferBase(gl.TRANSFORM_FEEDBACK_BUFFER, 2, destinationVBO[Particle.SPAWN_TIME]);
+            gl.bindBufferBase(gl.TRANSFORM_FEEDBACK_BUFFER, 3, destinationVBO[Particle.LIFE_TIME]);
+
+            // Turn off rasterization - we are not drawing
+            gl.enable(gl.RASTERIZER_DISCARD);
+
+            // Set uniforms
+            gl.uniform1f(emitTimeLocation, time);
+            gl.uniform2f(emitAccelerationLocation, 0.0, ACCELERATION);
+
+            // Emit particles using transform feedback
+            gl.beginTransformFeedback(gl.POINTS);
+            gl.drawArrays(gl.POINTS, 0, NUM_PARTICLES);
+            gl.endTransformFeedback();
+
+            // Create a sync object to ensure transform feedback results
+            // are completed before the draw that uses them.
+            // Sync object can't be reused so we create a new one for each synchronization operation.
+            // Not recommended to use gl.finish unless neccessary for synchronizing all your operation as it could severely reduce the GPU performance. Instead use sync object as you can wait for partial completion of the GL command stream.
+            emitSync = gl.fenceSync(gl.SYNC_GPU_COMMANDS_COMPLETE, 0);
+
+            // Restore state
+            gl.disable(gl.RASTERIZER_DISCARD);
+            gl.useProgram(null);
+            gl.bindBuffer(gl.ARRAY_BUFFER, null);
+            gl.bindBufferBase(gl.TRANSFORM_FEEDBACK_BUFFER, 0, null);
+            gl.bindBufferBase(gl.TRANSFORM_FEEDBACK_BUFFER, 1, null);
+            gl.bindBufferBase(gl.TRANSFORM_FEEDBACK_BUFFER, 2, null);
+            gl.bindBufferBase(gl.TRANSFORM_FEEDBACK_BUFFER, 3, null);
+            gl.bindTransformFeedback(gl.TRANSFORM_FEEDBACK, null);
+
+            // Ping pong the buffers
+            currentSourceIdx = (currentSourceIdx + 1) % 2;
+        }
+
+        function render() {
+            // Spawn particles
+            emitParticles();
+
+            time = Date.now();
+
+            // Block the GL server until transform feedback results are completed
+            gl.waitSync(emitSync, 0, gl.TIMEOUT_IGNORED);

```

Comment:
To be honest, they are not currently useful in WebGL 2.0. They were included for the sake of completeness; WebGL does not deviate arbitrarily from the OpenGL ES specifications. Also, if resource sharing between multiple WebGL contexts is (finally) added, then sync objects will immediately become necessary.